File: javav2/example_code/apigateway/src/test/java/APIGatewayTest.java
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import com.example.gateway.*;
import com.google.gson.Gson;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import software.amazon.awssdk.services.apigateway.ApiGatewayClient;
import org.junit.jupiter.api.*;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.secretsmanager.SecretsManagerClient;
import software.amazon.awssdk.services.secretsmanager.model.GetSecretValueRequest;
import software.amazon.awssdk.services.secretsmanager.model.GetSecretValueResponse;
import java.io.*;
import java.util.Random;
import static org.junit.jupiter.api.Assertions.*;

/**
 * To run these integration tests, you must set the required values
 * in the config.properties file or AWS Secrets Manager.
 */
@TestInstance(TestInstance.Lifecycle.PER_METHOD)
@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
public class APIGatewayTest {
    private static final Logger logger = LoggerFactory.getLogger(APIGatewayTest.class);
    private static ApiGatewayClient apiGateway;
    private static String restApiId = "";
    private static String resourceId = "";
    private static String httpMethod = "";
    private static String restApiName = "";
    private static String stageName = "";
    private static String newApiId = "";
    private static String deploymentId = "";

    @BeforeAll
    public static void setUp() throws IOException {
        Region region = Region.US_EAST_1;
        apiGateway = ApiGatewayClient.builder().region(region).build();
        Random random = new Random();
        int randomNum = random.nextInt((10000 - 1) + 1) + 1;
        // Get the values to run these tests from AWS Secrets Manager.
        Gson gson = new Gson();
        String json = getSecretValues();
        SecretValues values = gson.fromJson(json, SecretValues.class);
        restApiId = values.getRestApiId();
        httpMethod = values.getHttpMethod();
        restApiName = values.getRestApiName() + randomNum;
        stageName = values.getStageName();
    }

    @Test
    @Order(1)
    public void testCreateRestApi() {
        newApiId = CreateRestApi.createAPI(apiGateway, restApiId, restApiName);
        assertFalse(newApiId.isEmpty());
        logger.info("Test 1 passed");
    }

    @Test
    @Order(2)
    public void testCreateDeployment() {
        deploymentId = CreateDeployment.createNewDeployment(apiGateway, newApiId, stageName);
        assertFalse(deploymentId.isEmpty());
        logger.info("Test 2 passed");
    }

    @Test
    @Order(3)
    public void testGetDeployments() {
        assertDoesNotThrow(() -> GetDeployments.getAllDeployments(apiGateway, newApiId));
        logger.info("Test 3 passed");
    }

    @Test
    @Order(4)
    public void testGetStages() {
        assertDoesNotThrow(() -> GetStages.getAllStages(apiGateway, newApiId));
        logger.info("Test 4 passed");
    }

    @Test
    @Order(5)
    public void testDeleteRestApi() {
        assertDoesNotThrow(() -> DeleteRestApi.deleteAPI(apiGateway, newApiId));
        logger.info("Test 5 passed");
    }

    private static String getSecretValues() {
        SecretsManagerClient secretClient = SecretsManagerClient.builder()
                .region(Region.US_EAST_1)
                .build();
        String secretName = "test/apigateway";

        GetSecretValueRequest valueRequest = GetSecretValueRequest.builder()
                .secretId(secretName)
                .build();

        GetSecretValueResponse valueResponse = secretClient.getSecretValue(valueRequest);
        return valueResponse.secretString();
    }

    @Nested
    @DisplayName("A class used to get test values from test/apigateway (an AWS Secrets Manager secret)")
    class SecretValues {
        private String restApiId;
        private String restApiName;
        private String httpMethod;

        private String stageName;

        public String getRestApiId() {
            return restApiId;
        }

        public String getRestApiName() {
            return restApiName;
        }

        public String getHttpMethod() {
            return httpMethod;
        }

        public String getStageName() {
            return stageName;
        }
    }
}

File: kotlin/services/apigateway/src/test/kotlin/APIGatewayTest.kt
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import aws.sdk.kotlin.services.apigateway.ApiGatewayClient
import aws.sdk.kotlin.services.secretsmanager.SecretsManagerClient
import aws.sdk.kotlin.services.secretsmanager.model.GetSecretValueRequest
import com.google.gson.Gson
import com.kotlin.gateway.createAPI
import com.kotlin.gateway.deleteAPI
import com.kotlin.gateway.getAllDeployments
import com.kotlin.gateway.getAllStages
import kotlinx.coroutines.runBlocking
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.MethodOrderer.OrderAnnotation
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Order
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInstance
import org.junit.jupiter.api.TestMethodOrder
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.util.Random

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
@TestMethodOrder(OrderAnnotation::class)
class APIGatewayTest {
    private val logger: Logger = LoggerFactory.getLogger(APIGatewayTest::class.java)
    lateinit var apiGatewayClient: ApiGatewayClient
    private var restApiId = ""
    private var httpMethod = ""
    private var restApiName = ""
    private var stageName = ""
    private var newApiId = ""

    @BeforeAll
    fun setup() = runBlocking {
        apiGatewayClient = ApiGatewayClient { region = "us-east-1" }
        val random = Random()
        val randomNum = random.nextInt(10000 - 1 + 1) + 1
        val gson = Gson()
        val json: String = getSecretValues()
        val values: SecretValues = gson.fromJson(json, SecretValues::class.java)
        restApiId = values.restApiId.toString()
        httpMethod = values.httpMethod.toString()
        restApiName = values.restApiName.toString() + randomNum
        stageName = values.stageName.toString()
    }

    @Test
    @Order(1)
    fun createRestApiTest() = runBlocking {
        newApiId = createAPI(restApiId).toString()
        logger.info("Test 1 passed")
    }

    @Test
    @Order(2)
    fun getDeploymentsTest() = runBlocking {
        getAllDeployments(newApiId)
        logger.info("Test 2 passed")
    }

    @Test
    @Order(3)
    fun getAllStagesTest() = runBlocking {
        getAllStages(newApiId)
        logger.info("Test 3 passed")
    }

    @Test
    @Order(4)
    fun deleteRestApi() = runBlocking {
        deleteAPI(newApiId)
        logger.info("Test 4 passed")
    }

    private suspend fun getSecretValues(): String {
        val secretName = "test/apigateway"
        val valueRequest = GetSecretValueRequest {
            secretId = secretName
        }
        SecretsManagerClient {
            region = "us-east-1"
        }.use { secretClient ->
            val valueResponse = secretClient.getSecretValue(valueRequest)
            return valueResponse.secretString.toString()
        }
    }

    @Nested
    @DisplayName("A class used to get test values from test/apigateway (an AWS Secrets Manager secret)")
    internal class SecretValues {
        val restApiId: String? = null
        val restApiName: String? = null
        val httpMethod: String? = null
        val stageName: String? = null
    }
}
